### PR TITLE
Adjacent refactor

### LINKED ISSUE #744: Double Doors can only be pried open on one of the door tiles
As shown, attempting to pry open the door from the visible location fails. You must have access to the lower door section to pry it open.
77ea790372712ee1da61ba36d481d5fa

Reproduced this locally with no testmerges.

### PR DESCRIPTION
**Pull request by Rohesie**
Should fix #744 .
Untested yet, late night coding.
Used /tg/'s as a base, added multitle exceptions existing in our code and turned one recursive function into an iteration.

### CODE REVIEW

**Comment by DominikPanic** [comment]
What's the worst that could happen :ohgodohfuck:

File: code/_onclick/adjacent.dm [hunk 1]
```diff
@@ -25,50 +25,42 @@
...
-		return T0.ClickCross(get_dir(T0,src), border_only = 1) && src.ClickCross(get_dir(src,T0), border_only = 1, target_atom = target)
+		return T0.ClickCross(get_dir(T0,src), border_only = TRUE, target_atom = target, mover = mover) && ClickCross(get_dir(src,T0), border_only = 1, target_atom = target, mover = mover)
 
-	// Not orthagonal
-	var/in_dir = get_dir(neighbor,src) // eg. northwest (1+8)
-	var/d1 = in_dir&(in_dir-1)		// eg west		(1+8)&(8) = 8
-	var/d2 = in_dir - d1			// eg north		(1+8) - 8 = 1
+	// Diagonal case
+	var/in_dir = get_dir(T0,src) // eg. northwest (1+8) = 9 (00001001)
+	var/d1 = in_dir & 3		     // eg. north	  (1+8)&3 (0000 0011) = 1 (0000 0001)
```
Comment (Menshin) [inline code/_onclick/adjacent.dm]:
```suggestion
	var/d1 = NSCOMPONENT(in_dir)
```

File: code/_onclick/adjacent.dm [hunk 2]
```diff
@@ -25,50 +25,42 @@
...
+	var/d2 = in_dir & 12		 // eg. west	  (1+8)&12 (0000 1100) = 8 (0000 1000)
```
Comment (Menshin) [inline code/_onclick/adjacent.dm]:
```suggestion
	var/d2 = EWCOMPONENT(in_dir)
```

File: code/_onclick/adjacent.dm [hunk 3]
```diff
@@ -78,81 +70,79 @@ Quick adjacency (to turf):
...
-		if( !O.density || O == target_atom || O.throwpass) continue // throwpass is used for anything you can click through
+		if((mover && O.CanPass(mover,get_step(src, target_dir))) || (!mover && !O.density))
+			continue
+		if(O == target_atom || O == mover || O.throwpass) //check if there's a dense object present on the turf
+			continue // LETPASSTHROW is used for anything you can click through (or the firedoor special case, see above)
 
-		if( O.flags_atom & ON_BORDER) // windows have throwpass but are on border, check them first
-			if( O.dir & target_dir || O.dir&(O.dir-1) ) // full tile windows are just diagonals mechanically
+		if(O.flags_atom & ON_BORDER) // windows have throwpass but are on border, check them first
+			if(O.dir & target_dir || O.dir & (O.dir-1)) // full tile windows are just diagonals mechanically
```
Comment (Menshin) [inline code/_onclick/adjacent.dm]:
```suggestion
			if(O.dir & target_dir || ISDIAGONALDIR(O.dir)) // full tile windows are just diagonals mechanically
```

File: code/_onclick/adjacent.dm [hunk 4]
```diff
@@ -25,50 +25,42 @@
 	* If you are diagonally adjacent, ensure you can pass through at least one of the mutually adjacent square.
 		* Passing through in this case ignores anything with the throwpass flag, such as tables, racks, and morgue trays.
 */
-/turf/Adjacent(var/atom/neighbor, var/atom/target = null)
+/turf/Adjacent(atom/neighbor, atom/target = null, atom/movable/mover = null)
```
Comment (spookydonut) [inline code/_onclick/adjacent.dm]: they are null by default

Comment (Rohesie) [reply code/_onclick/adjacent.dm]: Yea, that's copied from /tg/ but I can clean it.

**Comment by Rohesie** [comment]
Not incredibly happy with the amount of snowflakery, but this should be better performant than the old code.

**Pull request merged**